8.0 stock quant merge

clonedagain opened the pull request:
Proposal to adopt OdooMRP's module to merge quants when they were split and it's not needed anymore (ie. reservation is undone).

To be done:
- [x] Add OCA as co-author
- [x] Use OCA's README template
- [x] Add automatic tests

oihane committed:
[ADD] New module <stock_quant_merge> closes #693
First upload, merge quants

oihane committed:
[MOD] Second step, quants merged when unreserving stock.move

oihane committed:
[MOD] Fixed some problems when there are quants with diferente histories

oihane committed:
[MOD] intermediate variable deleted

oihane committed:
[FIX] Unlinking quants fails with non-admin users, avoided. Closes #719

oihane committed:
[MOD] error when merging just deleted quants

oihane committed:
[MOD] Use browse record instead of ids

oihane committed:
[MOD] Write permission error solved

pedrobaeza committed:
[IMP] Expand AUTHORS

oihane committed:
Merge pull request #734 from oihane/8.0-quant_merge

[MOD] Write permission error solved

oihane committed:
[IMP] Fixed email address

oihane committed:
[FIX] <stock_quant_merge> Fixed cost calculation error

pedrobaeza committed:
Merge pull request #800 from oihane/8.0-quant_merge

[FIX] <stock_quant_merge> Fixed cost calculation error

pedrobaeza committed:
OCA Transbot updated translations from Transifex

Commit:
Add 'stock_quant_merge/' from commit '990e069455614dcd0781086e6af1f68ab2d021a0'

git-subtree-dir: stock_quant_merge
git-subtree-mainline: 3a0b6cfd13b0d63a47629a33103a5965fc22a141
git-subtree-split: 990e069455614dcd0781086e6af1f68ab2d021a0

clonedagain commented:
For you information @pedrobaeza @oihane @anajuaristi @lbellier

anajuaristi commented:
Stock_quant_merge is working in production for several customers so :+1: RTM from my side

rvalyi commented:
Looks good. It's something I always wanted.
But @clonedagain remarks should be implemented so we can merge it in the OCA.

Commit:
Adopt OCA conventions

Use OCA's the README template
Adapt manifest, add OCA as author
Update copyright notices

Commit:
Add an automatic test

clonedagain commented:
The code now adheres to the conventions and has automatic tests.
Reviewers wanted.

elicoidal commented:
Formally LGTM (no test, no technical check)

rvalyi commented on stock_quant_merge/models/stock.py:
```diff
@@ -0,0 +1,44 @@
+        for quant2merge in self:
+            if (quant2merge in pending_quants and
+                    not quant2merge.reservation_id):
+                quants = self.search(
```
could you please isolate this domain in a method that would take quant2merge as an argument and return the domain?
Some may need to override this domain. Example: in Brazil we will probably separate quants of the same product depending on if it comes from national origin or if it's imported. It's the same product in the MRP for instance. But for fiscal reports we cannot mix quants and we don't want to freak with lots for product origin. In this case we would need to override this domain, but may be there are other use cases.

oihane commented on stock_quant_merge/models/stock.py:
```diff
@@ -0,0 +1,44 @@
+
+    @api.multi
+    def merge_stock_quants(self):
+        pending_quants = self.filtered(lambda x: True)
```
Could you please explain why do you use pending_quants, wouldn't it been enough by processing those with `reservation_id` not set?

clonedagain replied:
Wait, _you_ wrote that didn't you? :-)
I think it avoids trying to merge a quant we just removed in a previous iteration.

oihane replied:
Yes, I did, but I don't remember why :worried: 
Maybe you should also start by doing the for only with those quants that do not have reservation...
I mean `for quant2merge in self.filtered(lambda x: not x.reservation_id)`

clonedagain replied:
I suppose that's a way to get a copy of the recordset?

Commit:
[IMP] make the merge domain modular

As explained by Raphaël Valyi in https://github.com/OCA/stock-logistics-warehouse/pull/94#discussion_r45978223 :
Some may need to override this domain. Example: in Brazil we will probably separate quants of the same product depending on if it comes from national origin or if it's imported. It's the same product in the MRP for instance. But for fiscal reports we cannot mix quants and we don't want to freak with lots for product origin. In this case we would need to override this domain, but may be there are other use cases.

clonedagain commented:
@rvalyi i've changed it as you suggested.

rvalyi commented:
@clonedagain thanks! and kudos for the work!

:+1:

Commit:
Explain intent of pending_quants and simplify loop condition

As suggested by @oihane

clonedagain commented on stock_quant_merge/models/stock.py:
```diff
@@ -0,0 +1,49 @@
+    @api.multi
+    def merge_stock_quants(self):
+        # Get a copy of the recorset
+        pending_quants = self.browse(self.ids)
```
@oihane is it clearer this way?

oihane replied:
why don't you change the _for_? If you get only the ones with reservation_id not set it will do less work, don't you think?

clonedagain replied:
The `for` is on `self` not `pending_quants`.

clonedagain replied:
For the main part I think it's OK because the module itself only calls this method on unreserved quants, but there may be corner cases if called through RPC or custom modules.

I'm sorry but it's a bit too subtle for me to research given the time I can spare on this.
Maybe you could propose a test to make sure it's OK?

oihane replied:
This is what I mean...

```
    @api.multi
    def merge_stock_quants(self):
        # Get a copy of the recorset
        pending_quants = self.browse(self.ids)
        for quant2merge in self.filtered(lambda x: not x.reservation_id):
            if quant2merge in pending_quants:
                quants = self.search(self._mergeable_domain())
                cont = 1
                cost = quant2merge.cost
                for quant in quants:
                    if (self._get_latest_move(quant2merge) ==
                            self._get_latest_move(quant)):
                        quant2merge.sudo().qty += quant.qty
                        cost += quant.cost
                        cont += 1
                        pending_quants -= quant
                        quant.sudo().unlink()
                quant2merge.sudo().cost = cost / cont
```

So this way you will only **for** those quants that have not been reserved yet.

@pedrobaeza, am I wrong?

clonedagain replied:
Ok changed as you suggested.

clonedagain commented:
@lbellier is it OK for you?

lbellier commented:
@clonedagain Yes, it's good for me.

Cedric-Pigeon commented:
:+1: (code review) Thanks !

oihane commented:
I see it ok :+1:
